File: driver-scala/src/main/scala/org/mongodb/scala/model/Updates.scala
/*
 * Copyright 2008-present MongoDB, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.mongodb.scala.model

import scala.collection.JavaConverters._

import com.mongodb.client.model.{ PushOptions => JPushOptions, Updates => JUpdates }

import org.mongodb.scala.bson.conversions.Bson

/**
 * A factory for document updates. A convenient way to use this class is to statically import all of its methods, which allows usage like:
 *
 * `collection.updateOne(eq("x", 1), set("x", 2))`
 *
 * @since 1.0
 */
object Updates {

  /**
   * Combine a list of updates into a single update.
   *
   * @param updates the list of updates
   * @return a combined update
   */
  def combine(updates: Bson*): Bson = JUpdates.combine(updates.asJava)

  /**
   * Creates an update that sets the value of the field with the given name to the given value.
   *
   * @param fieldName the non-null field name
   * @param value     the value
   * @tparam TItem   the value type
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/set/ \$set]]
   */
  def set[TItem](fieldName: String, value: TItem): Bson = JUpdates.set(fieldName, value)

  /**
   * Creates an update that sets the value for the document.
   *
   * @param value     the value
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/set/ \$set]]
   */
  def set(value: Bson): Bson = JUpdates.set(value)

  /**
   * Creates an update that deletes the field with the given name.
   *
   * @param fieldName the non-null field name
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/unset/ \$unset]]
   */
  def unset(fieldName: String): Bson = JUpdates.unset(fieldName)

  /**
   * Creates an update that sets the value of the field with the given name to the given value, but only if the update is an upsert that
   * results in an insert of a document.
   *
   * @param fieldName the non-null field name
   * @param value     the value
   * @tparam TItem   the value type
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/setOnInsert/ \$setOnInsert]]
   * @see UpdateOptions#upsert(boolean)
   */
  def setOnInsert[TItem](fieldName: String, value: TItem): Bson = JUpdates.setOnInsert(fieldName, value)

  /**
   * Creates an update that sets the values for the document, but only if the update is an upsert that results in an insert of a document.
   *
   * @param value     the value
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/setOnInsert/ \$setOnInsert]]
   * @see UpdateOptions#upsert(boolean)
   */
  def setOnInsert(value: Bson): Bson = JUpdates.setOnInsert(value)

  /**
   * Creates an update that renames a field.
   *
   * @param fieldName    the non-null field name
   * @param newFieldName the non-null new field name
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/rename/ \$rename]]
   */
  def rename(fieldName: String, newFieldName: String): Bson = JUpdates.rename(fieldName, newFieldName)

  /**
   * Creates an update that increments the value of the field with the given name by the given value.
   *
   * @param fieldName the non-null field name
   * @param number     the value
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/inc/ \$inc]]
   */
  def inc(fieldName: String, number: Number): Bson = JUpdates.inc(fieldName, number)

  /**
   * Creates an update that multiplies the value of the field with the given name by the given number.
   *
   * @param fieldName the non-null field name
   * @param number    the non-null number
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/mul/ \$mul]]
   */
  def mul(fieldName: String, number: Number): Bson = JUpdates.mul(fieldName, number)

  /**
   * Creates an update that sets the value of the field to the given value if the given value is less than the current value of the
   * field.
   *
   * @param fieldName the non-null field name
   * @param value     the value
   * @tparam TItem   the value type
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/min/ \$min]]
   */
  def min[TItem](fieldName: String, value: TItem): Bson = JUpdates.min(fieldName, value)

  /**
   * Creates an update that sets the value of the field to the given value if the given value is greater than the current value of the
   * field.
   *
   * @param fieldName the non-null field name
   * @param value     the value
   * @tparam TItem   the value type
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/min/ \$min]]
   */
  def max[TItem](fieldName: String, value: TItem): Bson = JUpdates.max(fieldName, value)

  /**
   * Creates an update that sets the value of the field to the current date as a BSON date.
   *
   * @param fieldName the non-null field name
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/currentDate/ \$currentDate]]
   * @see [[http://docs.mongodb.com/manual/reference/bson-types/#date Date]]
   */
  def currentDate(fieldName: String): Bson = JUpdates.currentDate(fieldName)

  /**
   * Creates an update that sets the value of the field to the current date as a BSON timestamp.
   *
   * @param fieldName the non-null field name
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/currentDate/ \$currentDate]]
   * @see [[http://docs.mongodb.com/manual/reference/bson-types/#document-bson-type-timestamp Timestamp]]
   */
  def currentTimestamp(fieldName: String): Bson = JUpdates.currentTimestamp(fieldName)

  /**
   * Creates an update that adds the given value to the array value of the field with the given name, unless the value is
   * already present, in which case it does nothing
   *
   * @param fieldName the non-null field name
   * @param value     the value
   * @tparam TItem   the value type
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/addToSet/ \$addToSet]]
   */
  def addToSet[TItem](fieldName: String, value: TItem): Bson = JUpdates.addToSet(fieldName, value)

  /**
   * Creates an update that adds each of the given values to the array value of the field with the given name, unless the value is
   * already present, in which case it does nothing
   *
   * @param fieldName the non-null field name
   * @param values     the values
   * @tparam TItem   the value type
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/addToSet/ \$addToSet]]
   */
  def addEachToSet[TItem](fieldName: String, values: TItem*): Bson = JUpdates.addEachToSet(fieldName, values.asJava)

  /**
   * Creates an update that adds the given value to the array value of the field with the given name.
   *
   * @param fieldName the non-null field name
   * @param value     the value
   * @tparam TItem   the value type
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/push/ \$push]]
   */
  def push[TItem](fieldName: String, value: TItem): Bson = JUpdates.push(fieldName, value)

  /**
   * Creates an update that adds each of the given values to the array value of the field with the given name.
   *
   * @param fieldName the non-null field name
   * @param values    the values
   * @tparam TItem   the value type
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/push/ \$push]]
   */
  def pushEach[TItem](fieldName: String, values: TItem*): Bson = JUpdates.pushEach(fieldName, values.asJava)

  /**
   * Creates an update that adds each of the given values to the array value of the field with the given name, applying the given
   * options for positioning the pushed values, and then slicing and/or sorting the array.
   *
   * @param fieldName the non-null field name
   * @param values    the values
   * @param options   the non-null push options
   * @tparam TItem   the value type
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/push/ \$push]]
   */
  def pushEach[TItem](fieldName: String, options: JPushOptions, values: TItem*): Bson =
    JUpdates.pushEach(fieldName, values.asJava, options)

  /**
   * Creates an update that removes all instances of the given value from the array value of the field with the given name.
   *
   * @param fieldName the non-null field name
   * @param value     the value
   * @tparam TItem   the value type
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/pull/ \$pull]]
   */
  def pull[TItem](fieldName: String, value: TItem): Bson = JUpdates.pull(fieldName, value)

  /**
   * Creates an update that removes from an array all elements that match the given filter.
   *
   * @param filter the query filter
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/pull/ \$pull]]
   */
  def pullByFilter(filter: Bson): Bson = JUpdates.pullByFilter(filter)

  /**
   * Creates an update that removes all instances of the given values from the array value of the field with the given name.
   *
   * @param fieldName the non-null field name
   * @param values    the values
   * @tparam TItem   the value type
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/pull/ \$pull]]
   */
  def pullAll[TItem](fieldName: String, values: TItem*): Bson = JUpdates.pullAll(fieldName, values.asJava)

  /**
   * Creates an update that pops the first element of an array that is the value of the field with the given name.
   *
   * @param fieldName the non-null field name
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/pop/ \$pop]]
   */
  def popFirst(fieldName: String): Bson = JUpdates.popFirst(fieldName)

  /**
   * Creates an update that pops the last element of an array that is the value of the field with the given name.
   *
   * @param fieldName the non-null field name
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/pop/ \$pop]]
   */
  def popLast(fieldName: String): Bson = JUpdates.popLast(fieldName)

  /**
   * Creates an update that performs a bitwise and between the given integer value and the integral value of the field with the given
   * name.
   *
   * @param fieldName the field name
   * @param value     the value
   * @return the update
   */
  def bitwiseAnd(fieldName: String, value: Int): Bson = JUpdates.bitwiseAnd(fieldName, value)

  /**
   * Creates an update that performs a bitwise and between the given long value and the integral value of the field with the given name.
   *
   * @param fieldName the field name
   * @param value     the value
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/bit/ \$bit]]
   */
  def bitwiseAnd(fieldName: String, value: Long): Bson = JUpdates.bitwiseAnd(fieldName, value)

  /**
   * Creates an update that performs a bitwise or between the given integer value and the integral value of the field with the given
   * name.
   *
   * @param fieldName the field name
   * @param value     the value
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/bit/ \$bit]]
   */
  def bitwiseOr(fieldName: String, value: Int): Bson = JUpdates.bitwiseOr(fieldName, value)

  /**
   * Creates an update that performs a bitwise or between the given long value and the integral value of the field with the given name.
   *
   * @param fieldName the field name
   * @param value     the value
   * @return the update
   * @see [[http://docs.mongodb.com/manual/reference/operator/update/bit/ \$bit]]
   */
  def bitwiseOr(fieldName: String, value: Long): Bson = JUpdates.bitwiseOr(fieldName, value)

  /**
   * Creates an update that performs a bitwise xor between the given integer value and the integral value of the field with the given
   * name.
   *
   * @param fieldName the field name
   * @param value     the value
   * @return the update
   */
  def bitwiseXor(fieldName: String, value: Int): Bson = JUpdates.bitwiseXor(fieldName, value)

  /**
   * Creates an update that performs a bitwise xor between the given long value and the integral value of the field with the given name.
   *
   * @param fieldName the field name
   * @param value     the value
   * @return the update
   */
  def bitwiseXor(fieldName: String, value: Long): Bson = JUpdates.bitwiseXor(fieldName, value)
}

File: driver-kotlin-extensions/src/main/kotlin/com/mongodb/kotlin/client/model/Updates.kt
/*
 * Copyright 2008-present MongoDB, Inc.
 * Copyright (C) 2016/2022 Litote
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @custom-license-header
 */
package com.mongodb.kotlin.client.model

import com.mongodb.client.model.PushOptions
import com.mongodb.client.model.UpdateOptions
import com.mongodb.client.model.Updates
import kotlin.internal.OnlyInputTypes
import kotlin.reflect.KProperty
import org.bson.conversions.Bson

/**
 * Updates extension methods to improve Kotlin interop
 *
 * @since 5.3
 */
@Suppress("TooManyFunctions")
public object Updates {

    /**
     * Creates an update that sets the value of the property to the given value.
     *
     * @param value the value
     * @param <T> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/set/ $set
     */
    @JvmSynthetic
    @JvmName("setExt")
    public infix fun <@OnlyInputTypes T> KProperty<T>.set(value: T?): Bson = Updates.set(path(), value)

    /**
     * Creates an update that sets the value of the property to the given value.
     *
     * @param property the data class property
     * @param value the value
     * @param <T> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/set/ $set
     */
    public fun <@OnlyInputTypes T> set(property: KProperty<T?>, value: T?): Bson = property.set(value)

    /**
     * Combine a list of updates into a single update.
     *
     * @param updates the list of updates
     * @return a combined update
     */
    public fun combine(vararg updates: Bson): Bson = Updates.combine(*updates)

    /**
     * Combine a list of updates into a single update.
     *
     * @param updates the list of updates
     * @return a combined update
     */
    public fun combine(updates: List<Bson>): Bson = Updates.combine(updates)

    /**
     * Creates an update that deletes the property with the given name.
     *
     * @param property the property
     * @return the update @mongodb.driver.manual reference/operator/update/unset/ $unset
     */
    public fun <T> unset(property: KProperty<T>): Bson = Updates.unset(property.path())

    /**
     * Creates an update that sets the value of the property to the given value, but only if the update is an upsert
     * that results in an insert of a document.
     *
     * @param value the value
     * @param <TItem> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/setOnInsert/ $setOnInsert
     * @see UpdateOptions#upsert(boolean)
     */
    @JvmSynthetic
    @JvmName("setOnInsertExt")
    public infix fun <@OnlyInputTypes T> KProperty<T?>.setOnInsert(value: T?): Bson = Updates.setOnInsert(path(), value)

    /**
     * Creates an update that sets the value of the property to the given value, but only if the update is an upsert
     * that results in an insert of a document.
     *
     * @param property the property
     * @param value the value
     * @param <TItem> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/setOnInsert/ $setOnInsert
     * @see UpdateOptions#upsert(boolean)
     */
    public fun <@OnlyInputTypes T> setOnInsert(property: KProperty<T?>, value: T?): Bson = property.setOnInsert(value)

    /**
     * Creates an update that renames a field.
     *
     * @param newProperty the new property
     * @return the update @mongodb.driver.manual reference/operator/update/rename/ $rename
     */
    @JvmSynthetic
    @JvmName("renameExt")
    public infix fun <@OnlyInputTypes T> KProperty<T?>.rename(newProperty: KProperty<T?>): Bson =
        Updates.rename(path(), newProperty.path())

    /**
     * Creates an update that renames a field.
     *
     * @param property the property
     * @param newProperty the new property
     * @return the update @mongodb.driver.manual reference/operator/update/rename/ $rename
     */
    public fun <@OnlyInputTypes T> rename(property: KProperty<T?>, newProperty: KProperty<T?>): Bson =
        property.rename(newProperty)

    /**
     * Creates an update that increments the value of the property by the given value.
     *
     * @param number the value
     * @return the update @mongodb.driver.manual reference/operator/update/inc/ $inc
     */
    @JvmSynthetic
    @JvmName("incExt")
    public infix fun <T : Number?> KProperty<T>.inc(number: Number): Bson = Updates.inc(path(), number)

    /**
     * Creates an update that increments the value of the property by the given value.
     *
     * @param property the property
     * @param number the value
     * @return the update @mongodb.driver.manual reference/operator/update/inc/ $inc
     */
    public fun <T : Number?> inc(property: KProperty<T>, number: Number): Bson = property.inc(number)

    /**
     * Creates an update that multiplies the value of the property by the given number.
     *
     * @param number the non-null number
     * @return the update @mongodb.driver.manual reference/operator/update/mul/ $mul
     */
    @JvmSynthetic
    @JvmName("mulExt")
    public infix fun <T : Number?> KProperty<T>.mul(number: Number): Bson = Updates.mul(path(), number)

    /**
     * Creates an update that multiplies the value of the property by the given number.
     *
     * @param property the property
     * @param number the non-null number
     * @return the update @mongodb.driver.manual reference/operator/update/mul/ $mul
     */
    public fun <T : Number?> mul(property: KProperty<T>, number: Number): Bson = property.mul(number)

    /**
     * Creates an update that sets the value of the property if the given value is less than the current value of the
     * property.
     *
     * @param value the value
     * @param <TItem> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/min/ $min
     */
    @JvmSynthetic
    @JvmName("minExt")
    public infix fun <@OnlyInputTypes T> KProperty<T>.min(value: T): Bson = Updates.min(path(), value)

    /**
     * Creates an update that sets the value of the property if the given value is less than the current value of the
     * property.
     *
     * @param property the property
     * @param value the value
     * @param <TItem> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/min/ $min
     */
    public fun <@OnlyInputTypes T> min(property: KProperty<T>, value: T): Bson = property.min(value)

    /**
     * Creates an update that sets the value of the property if the given value is greater than the current value of the
     * property.
     *
     * @param value the value
     * @param <TItem> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/min/ $min
     */
    @JvmSynthetic
    @JvmName("maxExt")
    public infix fun <@OnlyInputTypes T> KProperty<T>.max(value: T): Bson = Updates.max(path(), value)

    /**
     * Creates an update that sets the value of the property if the given value is greater than the current value of the
     * property.
     *
     * @param property the property
     * @param value the value
     * @param <TItem> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/min/ $min
     */
    public fun <@OnlyInputTypes T> max(property: KProperty<T>, value: T): Bson = property.max(value)

    /**
     * Creates an update that sets the value of the property to the current date as a BSON date.
     *
     * @param property the property
     * @return the update @mongodb.driver.manual reference/operator/update/currentDate/
     *   $currentDate @mongodb.driver.manual reference/bson-types/#date Date
     */
    public fun <T> currentDate(property: KProperty<T>): Bson = Updates.currentDate(property.path())

    /**
     * Creates an update that sets the value of the property to the current date as a BSON timestamp.
     *
     * @param property the property
     * @return the update @mongodb.driver.manual reference/operator/update/currentDate/
     *   $currentDate @mongodb.driver.manual reference/bson-types/#document-bson-type-timestamp Timestamp
     */
    public fun <T> currentTimestamp(property: KProperty<T>): Bson = Updates.currentTimestamp(property.path())

    /**
     * Creates an update that adds the given value to the array value of the property, unless the value is already
     * present, in which case it does nothing
     *
     * @param value the value
     * @param <TItem> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/addToSet/ $addToSet
     */
    @JvmSynthetic
    @JvmName("addToSetExt")
    public infix fun <@OnlyInputTypes T> KProperty<Iterable<T>?>.addToSet(value: T): Bson =
        Updates.addToSet(path(), value)

    /**
     * Creates an update that adds the given value to the array value of the property, unless the value is already
     * present, in which case it does nothing
     *
     * @param property the property
     * @param value the value
     * @param <TItem> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/addToSet/ $addToSet
     */
    public fun <@OnlyInputTypes T> addToSet(property: KProperty<Iterable<T>?>, value: T): Bson =
        property.addToSet(value)

    /**
     * Creates an update that adds each of the given values to the array value of the property, unless the value is
     * already present, in which case it does nothing
     *
     * @param property the property
     * @param values the values
     * @param <TItem> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/addToSet/ $addToSet
     */
    public fun <@OnlyInputTypes T> addEachToSet(property: KProperty<Iterable<T>?>, values: List<T>): Bson =
        Updates.addEachToSet(property.path(), values)

    /**
     * Creates an update that adds the given value to the array value of the property.
     *
     * @param property the property
     * @param value the value
     * @param <TItem> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/push/ $push
     */
    public fun <@OnlyInputTypes T> push(property: KProperty<Iterable<T>?>, value: T): Bson =
        Updates.push(property.path(), value)

    /**
     * Creates an update that adds each of the given values to the array value of the property, applying the given
     * options for positioning the pushed values, and then slicing and/or sorting the array.
     *
     * @param property the property
     * @param values the values
     * @param options the non-null push options
     * @param <TItem> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/push/ $push
     */
    public fun <@OnlyInputTypes T> pushEach(
        property: KProperty<Iterable<T>?>,
        values: List<T?>,
        options: PushOptions = PushOptions()
    ): Bson = Updates.pushEach(property.path(), values, options)

    /**
     * Creates an update that removes all instances of the given value from the array value of the property.
     *
     * @param value the value
     * @param <T> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/pull/ $pull
     */
    @JvmSynthetic
    @JvmName("pullExt")
    public infix fun <@OnlyInputTypes T> KProperty<Iterable<T?>?>.pull(value: T?): Bson = Updates.pull(path(), value)

    /**
     * Creates an update that removes all instances of the given value from the array value of the property.
     *
     * @param property the property
     * @param value the value
     * @param <T> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/pull/ $pull
     */
    public fun <@OnlyInputTypes T> pull(property: KProperty<Iterable<T?>?>, value: T?): Bson = property.pull(value)

    /**
     * Creates an update that removes all instances of the given value from the array value of the property.
     *
     * @param filter the value
     * @return the update @mongodb.driver.manual reference/operator/update/pull/ $pull
     */
    @JvmSynthetic
    @JvmName("pullByFilterExt")
    public infix fun KProperty<*>.pullByFilter(filter: Bson): Bson = Updates.pull(path(), filter)

    /**
     * Creates an update that removes all instances of the given value from the array value of the property.
     *
     * @param property the property
     * @param filter the value
     * @return the update @mongodb.driver.manual reference/operator/update/pull/ $pull
     */
    public fun pullByFilter(property: KProperty<*>, filter: Bson): Bson = property.pullByFilter(filter)

    /**
     * Creates an update that removes from an array all elements that match the given filter.
     *
     * @param filter the query filter
     * @return the update @mongodb.driver.manual reference/operator/update/pull/ $pull
     */
    public fun pullByFilter(filter: Bson): Bson = Updates.pullByFilter(filter)

    /**
     * Creates an update that removes all instances of the given values from the array value of the property.
     *
     * @param values the values
     * @param <TItem> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/pull/ $pull
     */
    @JvmSynthetic
    @JvmName("pullAllExt")
    public infix fun <@OnlyInputTypes T> KProperty<Iterable<T>?>.pullAll(values: List<T?>?): Bson =
        Updates.pullAll(path(), values ?: emptyList())

    /**
     * Creates an update that removes all instances of the given values from the array value of the property.
     *
     * @param property the property
     * @param values the values
     * @param <TItem> the value type
     * @return the update @mongodb.driver.manual reference/operator/update/pull/ $pull
     */
    public fun <@OnlyInputTypes T> pullAll(property: KProperty<Iterable<T>?>, values: List<T?>?): Bson =
        property.pullAll(values ?: emptyList())

    /**
     * Creates an update that pops the first element of an array that is the value of the property.
     *
     * @param property the property
     * @return the update @mongodb.driver.manual reference/operator/update/pop/ $pop
     */
    public fun <T> popFirst(property: KProperty<T>): Bson = Updates.popFirst(property.path())

    /**
     * Creates an update that pops the last element of an array that is the value of the property.
     *
     * @param property the property
     * @return the update @mongodb.driver.manual reference/operator/update/pop/ $pop
     */
    public fun <T> popLast(property: KProperty<T>): Bson = Updates.popLast(property.path())

    /**
     * Creates an update that performs a bitwise and between the given integer value and the integral value of the
     * property.
     *
     * @param value the value
     * @return the update
     */
    @JvmSynthetic
    @JvmName("bitwiseAndExt")
    public infix fun <T : Number?> KProperty<T>.bitwiseAnd(value: Int): Bson = Updates.bitwiseAnd(path(), value)

    /**
     * Creates an update that performs a bitwise and between the given integer value and the integral value of the
     * property.
     *
     * @param property the property
     * @param value the value
     * @return the update
     */
    public fun <T : Number?> bitwiseAnd(property: KProperty<T>, value: Int): Bson = property.bitwiseAnd(value)

    /**
     * Creates an update that performs a bitwise and between the given long value and the integral value of the
     * property.
     *
     * @param value the value
     * @return the update @mongodb.driver.manual reference/operator/update/bit/ $bit
     */
    @JvmSynthetic
    @JvmName("bitwiseAndExt")
    public infix fun <T : Number?> KProperty<T>.bitwiseAnd(value: Long): Bson = Updates.bitwiseAnd(path(), value)

    /**
     * Creates an update that performs a bitwise and between the given long value and the integral value of the
     * property.
     *
     * @param property the property
     * @param value the value
     * @return the update @mongodb.driver.manual reference/operator/update/bit/ $bit
     */
    public fun <T : Number?> bitwiseAnd(property: KProperty<T>, value: Long): Bson = property.bitwiseAnd(value)

    /**
     * Creates an update that performs a bitwise or between the given integer value and the integral value of the
     * property.
     *
     * @param value the value
     * @return the update @mongodb.driver.manual reference/operator/update/bit/ $bit
     */
    @JvmSynthetic
    @JvmName("bitwiseOrExt")
    public infix fun <T : Number?> KProperty<T>.bitwiseOr(value: Int): Bson = Updates.bitwiseOr(path(), value)

    /**
     * Creates an update that performs a bitwise or between the given integer value and the integral value of the
     * property.
     *
     * @param property the property
     * @param value the value
     * @return the update @mongodb.driver.manual reference/operator/update/bit/ $bit
     */
    public fun <T : Number?> bitwiseOr(property: KProperty<T>, value: Int): Bson =
        Updates.bitwiseOr(property.path(), value)

    /**
     * Creates an update that performs a bitwise or between the given long value and the integral value of the property.
     *
     * @param value the value
     * @return the update @mongodb.driver.manual reference/operator/update/bit/ $bit
     */
    @JvmSynthetic
    @JvmName("bitwiseOrExt")
    public infix fun <T : Number?> KProperty<T>.bitwiseOr(value: Long): Bson = Updates.bitwiseOr(path(), value)

    /**
     * Creates an update that performs a bitwise or between the given long value and the integral value of the property.
     *
     * @param property the property
     * @param value the value
     * @return the update @mongodb.driver.manual reference/operator/update/bit/ $bit
     */
    public fun <T : Number?> bitwiseOr(property: KProperty<T>, value: Long): Bson = property.bitwiseOr(value)

    /**
     * Creates an update that performs a bitwise xor between the given integer value and the integral value of the
     * property.
     *
     * @param value the value
     * @return the update
     */
    @JvmSynthetic
    @JvmName("bitwiseXorExt")
    public infix fun <T : Number?> KProperty<T>.bitwiseXor(value: Int): Bson = Updates.bitwiseXor(path(), value)

    /**
     * Creates an update that performs a bitwise xor between the given integer value and the integral value of the
     * property.
     *
     * @param property the property
     * @param value the value
     * @return the update
     */
    public fun <T : Number?> bitwiseXor(property: KProperty<T>, value: Int): Bson =
        Updates.bitwiseXor(property.path(), value)

    /**
     * Creates an update that performs a bitwise xor between the given long value and the integral value of the
     * property.
     *
     * @param value the value
     * @return the update
     */
    @JvmSynthetic
    @JvmName("addToSetExt")
    public infix fun <T : Number?> KProperty<T>.bitwiseXor(value: Long): Bson = Updates.bitwiseXor(path(), value)

    /**
     * Creates an update that performs a bitwise xor between the given long value and the integral value of the
     * property.
     *
     * @param property the property
     * @param value the value
     * @return the update
     */
    public fun <T : Number?> bitwiseXor(property: KProperty<T>, value: Long): Bson =
        Updates.bitwiseXor(property.path(), value)
}
